Feature/help widget

trixeenya opened the pull request.

trixeenya committed:
Implement HelpWidget widget

trixeenya committed:
Implement 4 realisations of HelpWidget

trixeenya committed:
Implement 4 realizations of widgets HelpWidget in storybook

trixeenya committed:
Fix text alignment in first HelpWidget

trixeenya committed:
Merge branch 'feature/components' into feature/help_widget

# Conflicts:
#	lib/src/components/card_rounded.dart

solid-yuriiprykhodko commented:
@trixeenya Please sync this with `master` -- the diff contains code from the previous PR (#2)

trixeenya committed:
Merge branch 'master' into feature/help_widget

solid-yuriiprykhodko commented on lib/src/components/card_rounded.dart:
```diff
@@ -18,15 +18,29 @@ class CardRounded extends StatelessWidget {
   /// A variable that sets the width of the card.
   final double? width;
 
+  /// A variable that sets the padding of the card.
```
```suggestion
  /// Padding for the card contents.
```

solid-yuriiprykhodko commented on lib/src/components/card_rounded.dart:
```diff
@@ -35,6 +49,9 @@ class CardRounded extends StatelessWidget {
     this.onClose,
+    this.closeButtonAlignment = Alignment.topRight,
+    this.customButtonIcon,
+    this.padding,
```
Setting a default in the constructor makes it clearer to the user, and makes for a more readable `build` method.
```suggestion
    this.padding = const EdgeInsets.symmetric(
              vertical: 12,
              horizontal: 12,
            ),
```

solid-yuriiprykhodko commented on lib/src/components/card_rounded.dart:
```diff
@@ -49,20 +66,33 @@ class CardRounded extends StatelessWidget {
+          child: _CloseButton(onTap: onClose, iconData: customButtonIcon),
+        );
+      } else {
+        throw Exception(
+          'Card supports only Alignment.topRight and Alignment.bottomRight',
+        );
```
This is more of a programmer error, not a runtime exception, is it?

Perhaps let's replace it with an `assert`, what do you think?

solid-yuriiprykhodko commented on lib/src/components/card_rounded.dart:
```diff
@@ -49,20 +66,33 @@ class CardRounded extends StatelessWidget {
-        top: offset,
-        child: _CloseButton(onTap: onClose),
-      );
+      if (closeButtonAlignment == Alignment.topRight) {
+        _closeButton = Positioned(
+          right: _defaultOffset,
+          top: _defaultOffset,
+          child: _CloseButton(onTap: onClose, iconData: customButtonIcon),
+        );
+      } else if (closeButtonAlignment == Alignment.bottomRight) {
+        _closeButton = Positioned(
+          right: _defaultOffset,
+          bottom: _defaultOffset,
+          child: _CloseButton(onTap: onClose, iconData: customButtonIcon),
+        );
```
The button is the same here, only the offset changes -- can we have only the logic-dependent code inside the conditional here, and keep the non-changing parts out?

solid-yuriiprykhodko commented on lib/src/components/card_rounded.dart:
```diff
@@ -92,9 +125,9 @@ class _CloseButton extends StatelessWidget {
-          Icons.close,
-          size: 10,
+        child: Icon(
+          iconData ?? Icons.close,
```
Let's move this default to the constructor too.

solid-yuriiprykhodko commented on lib/src/components/details_button.dart:
```diff
@@ -0,0 +1,46 @@
+                    color: hoverColor,
+                  ),
+            ),
+            Transform.rotate(
+              angle: arrowRotateAngle,
```
RotatedBox for simplicity?

solid-yuriiprykhodko commented on lib/src/components/flag_widget.dart:
```diff
@@ -0,0 +1,28 @@
+import 'package:flutter/material.dart';
+
+/// A widget with Ukraine flag colors
+class FlagWidget extends StatelessWidget {
+  /// Constructor
+  const FlagWidget({Key? key}) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    return Column(
+      children: const [
+        _FlagPart(color: Colors.blue),
+        _FlagPart(color: Colors.yellow),
+      ],
+    );
+  }
+}
+
+class _FlagPart extends StatelessWidget {
+  final Color color;
+
+  const _FlagPart({Key? key, required this.color}) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    return Expanded(child: Container(color: color));
+  }
+}
```
Loving the decomposition here, uncle Bob would be glad :)

solid-yuriiprykhodko commented on lib/src/components/hover_wrapper.dart:
```diff
@@ -34,11 +34,20 @@ class HoverWrapper extends StatefulWidget {
-      onExit: (_) => setState(() => _isHovered = false),
+      cursor: SystemMouseCursors.click,
+      onEnter: (_) {
+        setState(() => _isHovered = true);
+        _onHoverCallback(_isHovered);
+      },
+      onExit: (_) {
+        setState(() => _isHovered = false);
+        _onHoverCallback(_isHovered);
```
Move setState inside the `_onHoverCallback`?
Then `_onHoverCallback` -> `_onHoverToggle`?
```suggestion
        _onHoverToggle(true);
      },
      onExit: (_) {
        _onHoverToggle(false);
```

solid-yuriiprykhodko commented on lib/src/widgets/help_collection/first_help_widget.dart:
```diff
@@ -0,0 +1,53 @@
+      },
+      mainCardBuilder: (flag, title, desc, onDetails, onClose) {
+        return CardRounded(
+          key: const ValueKey(1),
```
Why this key here?

trixeenya replied:
It is just for AnimatedSwitcher, he is have problems with recognizing widgets, if you goona have 2 CardRounded widgets and trying to switch between them, AnimatedSwitcher will not recognize, that is 2 different widgets.

solid-yuriiprykhodko commented on lib/src/widgets/help_collection/first_help_widget.dart:
```diff
@@ -0,0 +1,53 @@
+          ),
+        );
+      },
+      mainCardBuilder: (flag, title, desc, onDetails, onClose) {
```
Oof, that's a lot of arguments for a builder.

solid-yuriiprykhodko replied:
I wonder if we could simplify this.

solid-yuriiprykhodko replied:
I think we can:

1. Remove the `flag` arg -- it's much better to just have the default flag as a separate Widget anyone can use.
2. Remove `title` and `desc` -- providing those from within the widget is unnecessary.
3. With the above, we're only left with the callbacks -- `onDetails` and `onClose`.
4. For the callbacks -- let's mimic `TextEditingController`. `HelpWidget` will have an argument -- `HelpWidgetController controller`, which will have a few methods like `close` and `showDetails`, which could be used to control a custom-made widget. This will make for a much better separation of concerns, and make the code easier to modify.
5. Extra -- let's make the transition animation builder external as well.

trixeenya committed:
Refactor HelpWidget

trixeenya committed:
Resolve requested issues

trixeenya committed:
Write doc for widgets

trixeenya committed:
Fix formatting in 2 files

trixeenya committed:
Add fade transition by side of slide transition to HelpWidget transition animation

solid-yuriiprykhodko commented on lib/src/components/card_rounded.dart:
```diff
@@ -35,11 +50,17 @@ class CardRounded extends StatelessWidget {
   @override
   Widget build(BuildContext context) {
-    final Widget _closeButton;
+    final Widget _positionedButton;
```
This is still a "close" button, is it not?

```suggestion
    final Widget _closeButton;
```

solid-yuriiprykhodko commented on lib/src/components/card_rounded.dart:
```diff
@@ -49,39 +70,47 @@ class CardRounded extends StatelessWidget {
+        bottom: closeButtonAlignment == Alignment.bottomRight
+            ? _defaultOffset
+            : null,
+        child: button,
```
Will this scale well if we want to add other alignments?

solid-yuriiprykhodko commented on lib/src/widgets/help_collection/fourth_help_widget.dart:
```diff
@@ -0,0 +1,84 @@
+                  ),
+                  const SizedBox(height: 15),
+                  DetailsButton(
+                    title: 'See what you can do',
```
Let's make this customizable as well.

trixeenya committed:
Fix close button hover bug

solid-yuriiprykhodko commented on lib/src/widgets/help_collection/second_help_widget.dart:
```diff
@@ -0,0 +1,72 @@
+                  const SizedBox(
+                    width: 220,
+                    child: Divider(height: 20),
+                  ),
```
The divider width shouldn't determine overall widget size, let's configure it appropriately.

```suggestion
            CardRounded(
              width: 250,
              onClose: controller.onClose,
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.center,
                mainAxisSize: MainAxisSize.min,
                children: [
                  const SizedBox(height: 25),
                  Text(
                    title,
                    style: Theme.of(context).textTheme.headline6?.copyWith(
                          fontWeight: FontWeight.w500,
                        ),
                  ),
                  const Divider(height: 20),
```

solid-yuriiprykhodko commented on lib/src/widgets/help_collection/second_help_widget.dart:
```diff
@@ -0,0 +1,72 @@
+            const Positioned(
+              top: -26,
+              left: 105,
+              // right: 0,
```
```suggestion
```

solid-yuriiprykhodko commented on lib/src/widgets/help_collection/third_help_widget.dart:
```diff
@@ -0,0 +1,129 @@
+  }
+}
+
+class _ActionButton extends StatefulWidget {
```
Let's split this out into a separate file.

Keeping several widgets in one file usually complicates things:
1. Moving widgets into different files later on makes the diff hard to read -- especially if some changes were made during the move.
2. Has the potential to create unnecessary connections, and impairs encapsulation and clear separation of concerns.

solid-yuriiprykhodko commented on lib/src/widgets/help_collection/third_help_widget.dart:
```diff
@@ -0,0 +1,129 @@
+      },
+      onTap: widget.onTap,
+      child: Container(
+        height: _ActionButton._height,
```
Let's move these constants into `_ActionButtonState`:
```suggestion
        height: _height,
```

solid-yuriiprykhodko commented on lib/src/widgets/help_widget.dart:
```diff
@@ -0,0 +1,228 @@
+  }
+}
+
+class _FlagCard extends StatelessWidget {
```
Please split out the widgets below this mark into separate files.

solid-yuriiprykhodko commented on lib/src/widgets/help_widget.dart:
```diff
@@ -0,0 +1,228 @@
+  /// A parameter that defines the direction of the animation.
+  final Axis axis;
+
+  /// It's a list of buttons that will be displayed in the options card.
+  final List<HelpOptionButton> options;
```
The options should be injected externally, not provided from within.

solid-yuriiprykhodko replied:
I mean that we shouldn't be providing those in a builder. It's OK to have them predefined, but that's a different layer of abstraction -- they belong with the 4 default widget implementations, not with the base widgets.

trixeenya committed:
Change design of widgets to pixel perfect sign

trixeenya committed:
Merge remote-tracking branch 'origin/bugfix/close-button' into feature/help_widget

# Conflicts:
#	lib/src/components/buttons/help_option_button.dart
#	lib/src/widgets/links_card_widget.dart

trixeenya committed:
Fix closeButton var naming

trixeenya committed:
Solve requested issues

trixeenya committed:
Fix linter info

trixeenya committed:
Fix linter formatting

trixeenya committed:
Refactor HelpWidget switch logic

trixeenya committed:
Fix collapsed view in 3 widgets

solid-yuriiprykhodko commented on lib/src/theme/text_themes.dart:
```diff
@@ -0,0 +1,29 @@
+    // bodyText2: mainFont.copyWith(fontSize: 16),
+    // subtitle1: mainFont.copyWith(fontSize: 15),
+    // subtitle2: mainFont.copyWith(fontSize: 13),
+    // caption: mainFont.copyWith(fontSize: 12),
```
```suggestion
```

solid-yuriiprykhodko commented on lib/src/helpers/help_widget_view.dart:
```diff
@@ -0,0 +1,17 @@
+///
+/// Helps to show different views
+enum HelpWidgetView {
+  /// It's a constant value of type [HelpWidgetView] that
+  /// has the value `collapsed`.
```
This description is not really useful, is it?

How about we explain what this enum case is supposed to describe:
```suggestion
  /// A small, non-intrusive thumbnail view of the Widget.
```

solid-yuriiprykhodko commented on lib/src/helpers/help_widget_view.dart:
```diff
@@ -0,0 +1,17 @@
+/// It's defining a new type called [HelpWidgetView] with three possible values:
+/// `collapsed`, `main`, and `options`.
```
It is an enum by definition -- we don't need to state the obvious here. Besides -- it's probably not a good idea to list all the values in a doc comment -- it easily get out of date and confuses the reader.

```suggestion
/// A listing of available views of a single HelpUkraineWidget.
```

solid-yuriiprykhodko commented on lib/src/widgets/help_collection/first_help_widget.dart:
```diff
@@ -0,0 +1,63 @@
+import 'package:flutter/material.dart';
+import 'package:help_ukraine_widget/help_ukraine_widget.dart';
+
+/// [FirstHelpWidget] is modification of a [HelpWidget] described as first
+/// type.
```
Let's give some useful info here:
```suggestion
/// This is a variation of the widget that is shaped as a wide and short rectange.
///
/// Best suited for layouts where you need to save some vertical space.
```

solid-yuriiprykhodko replied:
This could also be a way to name them. Not sure calling the variations `First` and `Second` is descriptive. Instead we could do `WideShortHelpUkraineWidget`. 

What do you think? Maybe you have some ideas for this?

solid-yuriiprykhodko commented on lib/src/widgets/help_collection/fourth_help_widget.dart:
```diff
@@ -0,0 +1,94 @@
+
+  @override
+  Widget build(BuildContext context) {
+    final _controller = HelpWidgetViewController(HelpWidgetView.collapsed);
```
Are you sure you want to create a new controller for each rebuild?

solid-yuriiprykhodko commented on lib/src/widgets/help_collection/second_help_widget.dart:
```diff
@@ -0,0 +1,85 @@
+  /// A title of a widget.
+  final String title;
+
+  /// A description for the button that will be shown on the main card.
+  final String detailsButtonDescription;
```
```suggestion
  /// Title of the clickable text prompting the user to see the help options.
  final String detailsButtonDescription;
```

solid-yuriiprykhodko commented on lib/src/widgets/help_widget.dart:
```diff
@@ -0,0 +1,177 @@
+/// can be expanded, and a widget that can be displayed when
+/// the widget is expanded.
+class HelpWidget extends StatefulWidget {
+  /// A parameter that defines the direction of the animation.
+  final Axis axis;
```
```suggestion
  /// A parameter that defines the direction of the view transition animation.
  final Axis viewTransitionAxis;
```

solid-yuriiprykhodko commented on lib/src/widgets/help_widget.dart:
```diff
@@ -0,0 +1,177 @@
+    required HelpWidgetViewController controller,
+    this.axis = Axis.vertical,
+    this.constraints,
+    // required this.controller,
```
```suggestion
```

solid-yuriiprykhodko commented on lib/src/widgets/help_widget.dart:
```diff
@@ -0,0 +1,177 @@
+
+  final HelpWidgetViewController _controller;
+
+  /// It's a parameter that defines the maximum width of the widget.
+  final BoxConstraints? constraints;
```
```suggestion
  /// It's a parameter that defines the maximum size of the widget.
  final BoxConstraints? maxSize;
```

solid-yuriiprykhodko reviewed (changes requested):
This is customer-facing API -- let's make sure everything's squeaky clean and provide good documentation.

solid-yuriiprykhodko commented on lib/src/widgets/links_card_widget.dart:
```diff
@@ -0,0 +1,89 @@
+  }
+}
+
+class _LinksCardButton extends StatefulWidget {
```
Please move this to a separate file. We shouldn't really have 2 Widgets in one file -- anywhere.

solid-yuriiprykhodko committed:
Make font closer to original, fix trailing padding

dazzlemon committed:
renamed help_widgets, changed some of the doc comments to more descriptive ones

dazzlemon committed:
renamed help_widgets in the files where they are used

dazzlemon committed:
split help_widget.dart into separate files (one widget per file), added comments for previously private classes

dazzlemon committed:
split links_card_widget and links_card_button into separate files

dazzlemon committed:
split card_rounded and close_button

illia-romanenko committed:
Fix discussed issues with fonts in 1 place/requires other places to be rechecked.

dazzlemon committed:
moved _controller outside build method for HelpWidget variants, constructors are non const now

dazzlemon committed:
changed LinksCardWidget to StatelessWidget

dazzlemon committed:
changed HelpWidget to extend StatelessWidget

dazzlemon committed:
refactored view selection in HelpWidget.build

solid-yuriiprykhodko committed:
Merge branch 'first-widget-font-and-padding-fix' into feature/help_widget

solid-yuriiprykhodko committed:
Fix merge issue

solid-yuriiprykhodko committed:
Remove widget variations 2, 3 and 4

dazzlemon committed:
rm HelpWidgetSlideTransition

dazzlemon committed:
rm unused import

dazzlemon committed:
fix formatting

solid-yuriiprykhodko commented on lib/src/widgets/help_widget/help_widget_animation_builder.dart:
```diff
@@ -0,0 +1,82 @@
+
+    final fadeAnimation = _fadeAnimation(animation);
+
+    const padding = EdgeInsets.all(12.0);
+    final transition = FadeTransition(
+      opacity: fadeAnimation,
+      child: SlideTransition(
+        position: child.key == const ValueKey(0) || !isPositiveDirection
```
Looks like a bug. We no longer use integers to identify views, it's now an enum:

```suggestion
    const padding = EdgeInsets.all(12.0);
    final transitioningToCollapsed = child.key == const ValueKey(HelpWidgetView.collapsed);
    final transition = FadeTransition(
      opacity: fadeAnimation,
      child: SlideTransition(
        position: transitioningToCollapsed || !isPositiveDirection
```

solid-yuriiprykhodko reviewed (changes requested):
Just one last thing and we're good to go.

dazzlemon committed:
bugfix

HelpWidgetAnimationBuilder.child.key is always ValueKey<HelpWidgetView>, old comparsion was used when it was ValueKey<int>

Co-authored-by: Yuri Prykhodko <40931732+solid-yuriiprykhodko@users.noreply.github.com>

dazzlemon committed:
added missing import

dazzlemon committed:
fix formating

dazzlemon committed:
refactored unsafe comparison of Widget.key

dazzlemon committed:
placed line to correct place

dazzlemon committed:
formatting

solid-yuriiprykhodko commented on lib/src/widgets/help_widget/help_widget.dart:
```diff
@@ -0,0 +1,77 @@
+      child: ValueListenableBuilder<HelpWidgetView>(
+        valueListenable: _controller,
+        builder: (context, view, _) {
+          final isCollapsedView = view == HelpWidgetView.collapsed;
+
+          return GestureDetector(
+            onTap: () {
+              if (isCollapsedView) {
+                _controller.showMainView();
+              }
+            },
+            child: HelpWidgetAnimationBuilder(
+              transitioningToCollapsed: isCollapsedView,
+              axis: axis,
+              isPositiveDirection: _controller.isPositiveDirection,
```
Let's improve the naming and simplify:

```suggestion
          final isCurrentViewCollapsed = view == HelpWidgetView.collapsed;

          return GestureDetector(
            onTap: () {
              if (isCurrentViewCollapsed) {
                _controller.showMainView();
              }
            },
            child: HelpWidgetAnimationBuilder(
              axis: axis,
              transitionForward: isCurrentViewCollapsed || _controller.transitionForward,
```

solid-yuriiprykhodko replied:
But if we're assigning `transitionForward` due to logic inside the controller -- we should probably move it there too.

dazzlemon committed:
combine two fields into one in HelpWidgetAnimationBuilder

dazzlemon committed:
format

dazzlemon committed:
doc comments

solid-yuriiprykhodko committed:
Simplify transition picking logic

solid-yuriiprykhodko committed:
Remove commented out code

solid-yuriiprykhodko committed:
Remove collapsed view hit detection for base HelpWidget

solid-yuriiprykhodko committed:
Cleanup

solid-yuriiprykhodko reviewed (approved):
LGTM!
